src/libAtomVM/term.c

```diff
@@ -499,14 +513,52 @@ TermCompareResult term_compare(term t, term other, TermCompareOpts opts, GlobalC
             break;
 
         } else if (term_is_reference(t) && term_is_reference(other)) {
-            int64_t t_ticks = term_to_ref_ticks(t);
-            int64_t other_ticks = term_to_ref_ticks(other);
-            if (t_ticks == other_ticks) {
-                CMP_POP_AND_CONTINUE();
+            if (!term_is_external(t) && !term_is_external(other)) {
+                int64_t t_ticks = term_to_ref_ticks(t);
+                int64_t other_ticks = term_to_ref_ticks(other);
+                if (t_ticks == other_ticks) {
+                    CMP_POP_AND_CONTINUE();
+                } else {
+                    result = (t_ticks > other_ticks) ? TermGreaterThan : TermLessThan;
+                    break;
+                }
+            } else if (term_is_external(t) && term_is_external(other)) {
+                term node = term_get_external_node(t);
+                term other_node = term_get_external_node(other);
+                if (node == other_node) {
+                    uint32_t creation = term_get_external_node_creation(t);
+                    uint32_t other_creation = term_get_external_node_creation(other);
+                    if (creation == other_creation) {
+                        uint32_t len = term_get_external_reference_len(t);
+                        uint32_t other_len = term_get_external_reference_len(other);
+                        if (len == other_len) {
+                            uint32_t data[len];
+                            uint32_t other_data[len];
+                            term_get_external_reference_words(t, data);
+                            term_get_external_reference_words(other, other_data);
+                            int cmp = memcmp(data, other_data, len * sizeof(uint32_t));
+                            if (cmp == 0) {
+                                CMP_POP_AND_CONTINUE();
+                            } else {
+                                result = (cmp > 0) ? TermGreaterThan : TermLessThan;
+                                break;
+                            }
+                        } else {
+                            result = (len > other_len) ? TermGreaterThan : TermLessThan;
+                            break;
+                        }
+                    } else {
+                        result = (creation > other_creation) ? TermGreaterThan : TermLessThan;
+                        break;
+                    }
+                } else {
+                    result = (node > other_node) ? TermGreaterThan : TermLessThan;
+                    break;
+                }
             } else {
-                result = (t_ticks > other_ticks) ? TermGreaterThan : TermLessThan;
-                break;
+                result = term_is_external(t) ? TermGreaterThan : TermLessThan;
```

Comment:
If all external references are greater than local ones, we might resort to the general comparison for "different types" that makes use of `term_type_to_index`:
`result = (term_type_to_index(t) > term_type_to_index(other)) ? TermGreaterThan : TermLessThan`